design/proposals/modular-components.md

```diff
@@ -0,0 +1,234 @@
+   number of environments grow. We needn't store multiple copies of the same
+   API version on disk. The Kubernetes API version that each environment uses
+   can be recorded in the environment specification file. The metadata files
+   can be locally cached once per k8s API version in `lib`.
+
+3. A component should not be uniquely defined by it's base file name.
+   * The existing implementation implies that there cannot be a `foo.jsonnet`
+   and a `foo.yaml`, since the base file name is shared.
+
+4. Handle environment specifications in `app.yaml` and retire `spec.json`.
+   * Multiple files and areas for configuration can cause confusions for users.
+   In order to keep configurations simple, I propose that all ksonnet-specific
+   _app-level_ configurations reside in the `app.yaml`.
+
+`app.yaml` would add an _environments_ section as follows:
+
+```
+apiVersion: 0.0.1
+kind: ksonnet.io/app
+name: config
+version: 0.0.1
+registries:
+  incubator:
+    gitVersion:
+      commitSha: ea3408d44c2d8ea4d321364e5533d5c60e74bce0
+      refSpec: master
+    protocol: github
+    uri: github.com/ksonnet/parts/tree/master/incubator
+environments:
+  dev:
+    k8sVersion: 1.9.0
+    server: https://my-dev-server.com
+    namespaces:
+      - dev-foo
+      - dev-bar
+    targets:
+      - auth
+      - dev
+  staging:
+    k8sVersion: 1.8.0
+    server: https://my-staging-server.com
+    namespaces:
+      - staging
+    targets:
+      - redis.jsonnet
+      - auth
+      - prod
+```
+
+The _environments_ section operates similarly to the existing `spec.json` with
+the additions of:
+
+1. Support for multiple environment namespaces. Deployments will happen across
+   all the namespaces for that environment. If a user wishes to have separate
+   deployments to separate namespaces, these deployment groups should be
+   modeled as two separate environments.
+2. The introduction of _targets_. _Targets_ are the component targets that an
+   environment wishes to include in a deployment. _Target_ definitions are
+   equivalent to the directory or file name of the component.
+3. The introduction of _k8sVersion_, the Kubernetes version the environment
+   server is running on.
+
+Following the above `app.yaml` example, the users directory tree may look
+something like the following _(hidden files / directories are excluded)_:
+
+```
+my-ksonnet-app
+├── app.yaml
+├── components
+│   ├── auth
+│   │   ├── ca-secret.jsonnet
+│   │   └── tls-certificate.jsonnet
+│   ├── dev
+│   │   └── memcached.jsonnet
+│   ├── params.libsonnet
+│   ├── prod
+│   │   └── database
+│   │       ├── db-secret.jsonnet
+│   │       ├── db-service.yaml
+│   │       └── db-stateful.jsonnet
+│   └── redis.jsonnet
+├── environments
+│   ├── dev
+│   │   └── params.libsonnet
+│   └── prod
+│       └── params.libsonnet
+├── lib
+└── vendor
+```
+
+_Note_: There is no correlation between the environment _dev_ and the
```

Comment:
This might be bike shedding, but I feel like the figuring out best practices around environments/targets is good to earlier than later, and thus may effect the design so I wanted to comment on the example.

While I understand the design doesn't preclude this, it feels like the example here is an anti-pattern. 

I think it would make more sense to put the components at a level up (eg: `components/database` and `components/memcached.jsonnet), and then if you wanted `prod` or `dev`, it would make more sense to have a `dev.jsonnet` and `prod.jsonnet` which imports the env specific components and calls them directly, passing params through.

So then the `staging` environment from the example would adjust it's targets to this:
```
staging:
   targets:
     - prod.jsonnet
```

This wouldn't preclude listing more targets obviously as it was originally, but fits more into the documented model of a component, where a component could be something as complex as an entire environment. It also makes all the individual components more or less environment agnostic (is the db really prod only? is memcached, really dev only?). The alternative is that this `prod.jsonnet` or `dev.jsonnet` could live outside the `components` and be referenced as an alternative target.

Comparing this Chef in config management land, this is basically like `components==recipes`, `environments==environments/app.yaml`, and `roles`. are effectively a the `targets` in app.yaml. The idea in Chef is that environments can refer to individual recipes (components), or roles, and roles provide a way to group various recipes (components) together. I feel like we want a similar concept here, to group components either by role, environment, etc, which can be referred to as a target, as an alternative to listing all targets in app.yaml.

Reply:
This is great, thanks for all your feedback.

I agree that the example in the proposal might have poorly highlighted how a real scenario might be modeled. I wanted to present that if a production and a dev environment had a large set of configurations that differ greatly from each other (_i.e._ `database` in this scenario), it could be modeled in their own directory. For example,

```
├── prod
│   └── database
│       └── a-config.jsonnet
├── devel
│   └── database
│       └── a-completely-different-config.jsonnet
```

```
prod:
  targets:
    - prod
dev:
  targets:
    - devel
```

If `database` has the ability to be shared easily across multiple environments, then it makes sense to have `database` at the top-level, as you mentioned.

I believe your proposal of `prod.jsonnet` and `dev.jsonnet` is similar to how `main.jsonnet` for each environment operates today. We're still evaluating whether they exist well alongside this _targets_ model.